File: src/test/java/io/primeaspect/calculator/controller/StatisticsControllerTest.java
package io.primeaspect.calculator.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.primeaspect.calculator.dto.response.ExpressionListResponse;
import io.primeaspect.calculator.model.Expression;
import io.primeaspect.calculator.model.Number;
import io.primeaspect.calculator.service.StatisticsService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(controllers = StatisticsController.class)
public class StatisticsControllerTest {
    @Autowired
    private MockMvc mvc;
    @MockBean
    private StatisticsService service;
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Test
    public void findAllByDateTest() throws Exception {
        LocalDate date = LocalDate.of(2017, 12, 15);
        Expression response = new Expression(date, "(-78+9-(9/4.5))^2", 5401.0);
        List<Expression> responseList = new ArrayList<>();
        responseList.add(response);
        ExpressionListResponse expected = new ExpressionListResponse(responseList);

        when(service.findAllByDate(date.toString())).thenReturn(expected);

        mvc.perform(get("/statistics/date?date=2017-12-15"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.list").isArray())
                .andExpect(jsonPath("$.list", hasSize(1)))
                .andExpect(jsonPath("$.list[0].date", is(response.getDate().toString())))
                .andExpect(jsonPath("$.list[0].expression", is(response.getExpression())))
                .andExpect(jsonPath("$.list[0].solution", is(response.getSolution())));

        verify(service).findAllByDate(response.getDate().toString());
    }

    @Test
    public void findAllByOperationTest() throws Exception {
        LocalDate date = LocalDate.of(2017, 12, 15);
        Expression response = new Expression(date, "(-78+9-(9/4.5))^2", 5401.0);
        List<Expression> responseList = new ArrayList<>();
        responseList.add(response);
        ExpressionListResponse expected = new ExpressionListResponse(responseList);

        when(service.findAllByOperation("^")).thenReturn(expected);

        mvc.perform(get("/statistics/operation?operation=^"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.list").isArray())
                .andExpect(jsonPath("$.list", hasSize(1)))
                .andExpect(jsonPath("$.list[0].date", is(response.getDate().toString())))
                .andExpect(jsonPath("$.list[0].expression", is(response.getExpression())))
                .andExpect(jsonPath("$.list[0].solution", is(response.getSolution())));

        verify(service).findAllByOperation("^");
    }

    @Test
    public void getCountByDateTest() throws Exception {
        LocalDate date = LocalDate.of(2017, 12, 15);
        when(service.getCountByDate(date.toString())).thenReturn(5);

        mvc.perform(get("/statistics/count/date?date=2017-12-15"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", is(5)));

        verify(service).getCountByDate(date.toString());
    }

    @Test
    public void getCountByOperationTest() throws Exception {
        when(service.getCountByOperation("+")).thenReturn(5);

        mvc.perform(get("/statistics/count/operation?operation=+"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", is(5)));

        verify(service).getCountByOperation("+");
    }

    @Test
    public void getPopularNumberTest() throws Exception {
        Number expected = new Number(10, 5);

        when(service.getPopularNumber()).thenReturn(expected);

        mvc.perform(get("/statistics/popular"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.value", is(expected.getValue())))
                .andExpect(jsonPath("$.count", is(expected.getCount())));

        verify(service).getPopularNumber();
    }
}

File: src/test/kotlin/io/primeaspect/calculator/controller/StatisticsControllerTest.kt
package io.primeaspect.calculator.controller

import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import io.primeaspect.calculator.dto.response.ExpressionListResponse
import io.primeaspect.calculator.dto.response.NumberResponse
import io.primeaspect.calculator.model.Expression
import io.primeaspect.calculator.service.StatisticsService
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.hasSize
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import java.time.LocalDate

@WebMvcTest(controllers = [StatisticsController::class])
class StatisticsControllerTest {
    @Autowired
    private lateinit var mvc: MockMvc

    @MockBean
    private lateinit var service: StatisticsService

    @Test
    fun `Returns success status and a list of all expressions filtered by date when a GET request is sent`() {
        val date = LocalDate.of(2017, 12, 15)
        val response = Expression(date = date, expression = "(-78+9-(9/4.5))^2", solution = 5401.0)
        val responseList: List<Expression> = listOf(response)
        val expected = ExpressionListResponse(responseList)

        whenever(service.findAllByDate(date)).thenReturn(expected)

        mvc.perform(MockMvcRequestBuilders.get("/statistics/date?date=2017-12-15"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.list").isArray)
                .andExpect(MockMvcResultMatchers.jsonPath("$.list", hasSize<Any>(1)))
                .andExpect(MockMvcResultMatchers.jsonPath("$.list[0].date", `is`(response.date.toString())))
                .andExpect(MockMvcResultMatchers.jsonPath("$.list[0].expression", `is`(response.expression)))
                .andExpect(MockMvcResultMatchers.jsonPath("$.list[0].solution", `is`(response.solution)))
        verify(service).findAllByDate(response.date)
    }

    @Test
    fun `Returns success status and a list of all expressions filtered by operation when a GET request is sent`() {
        val date = LocalDate.of(2017, 12, 15)
        val response = Expression(date = date, expression = "(-78+9-(9/4.5))^2", solution = 5401.0)
        val responseList: List<Expression> = listOf(response)
        val expected = ExpressionListResponse(responseList)

        whenever(service.findAllByOperation("^")).thenReturn(expected)

        mvc.perform(MockMvcRequestBuilders.get("/statistics/operation?operation=^"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.list").isArray)
                .andExpect(MockMvcResultMatchers.jsonPath("$.list", hasSize<Any>(1)))
                .andExpect(MockMvcResultMatchers.jsonPath("$.list[0].date", `is`(response.date.toString())))
                .andExpect(MockMvcResultMatchers.jsonPath("$.list[0].expression", `is`(response.expression)))
                .andExpect(MockMvcResultMatchers.jsonPath("$.list[0].solution", `is`(response.solution)))
        verify(service).findAllByOperation("^")
    }

    @Test
    fun `Returns success status and count of all expressions filtered by date when a GET request is sent`() {
        val date = LocalDate.of(2017, 12, 15)

        whenever(service.getCountByDate(date)).thenReturn(5)

        mvc.perform(MockMvcRequestBuilders.get("/statistics/count/date?date=2017-12-15"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$", `is`(5)))
        verify(service).getCountByDate(date)
    }

    @Test
    fun `Returns success status and count of all expressions filtered by operation when a GET request is sent`() {
        whenever(service.getCountByOperation("+")).thenReturn(5)

        mvc.perform(MockMvcRequestBuilders.get("/statistics/count/operation?operation=+"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$", `is`(5)))
        verify(service).getCountByOperation("+")
    }

    @Test
    fun `Returns success status and the most popular number when a GET request is sent`() {
        val expected = NumberResponse(value = 10.0, count = 5)

        whenever(service.getPopularNumber()).thenReturn(expected)

        mvc.perform(MockMvcRequestBuilders.get("/statistics/popular"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.value", `is`(expected.value)))
                .andExpect(MockMvcResultMatchers.jsonPath("$.count", `is`(expected.count)))
        verify(service).getPopularNumber()
    }

    @Test
    fun `Returns success status and the most popular number by date when a GET request is sent`() {
        val date = LocalDate.of(2017, 12, 15)
        val expected = NumberResponse(value = 10.0, count = 5)

        whenever(service.getPopularNumberByDate(date)).thenReturn(expected)

        mvc.perform(MockMvcRequestBuilders.get("/statistics/popular/date?date=2017-12-15"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.value", `is`(expected.value)))
                .andExpect(MockMvcResultMatchers.jsonPath("$.count", `is`(expected.count)))
        verify(service).getPopularNumberByDate(date)
    }
}